src/WriterImpl.cpp

```diff
@@ -0,0 +1,970 @@
+   {
+      int64_t pos = -1;
+
+      if ( data3DHeader.guid.empty() )
+      {
+         data3DHeader.guid = generateRandomGUID();
+      }
+
+      StructureNode scan = StructureNode( imf_ );
+      data3D_.append( scan );
+      pos = data3D_.childCount() - 1;
+
+      scan.set( "guid", StringNode( imf_, data3DHeader.guid ) ); // required
+
+      if ( !data3DHeader.name.empty() )
+         scan.set( "name", StringNode( imf_, data3DHeader.name ) );
+
+      if ( !data3DHeader.description.empty() )
+         scan.set( "description", StringNode( imf_, data3DHeader.description ) );
+
+      if ( data3DHeader.originalGuids.size() > 0 )
+      {
+         scan.set( "originalGuids", VectorNode( imf_ ) );
+         VectorNode originalGuids( scan.get( "originalGuids" ) );
+         int i;
+         for ( i = 0; i < (int)data3DHeader.originalGuids.size(); i++ )
+            originalGuids.append( StringNode( imf_, data3DHeader.originalGuids[i] ) );
+      }
+
+      // Add various sensor and version strings to scan.
+      // Path names: "/data3D/0/sensorVendor", etc...
+      if ( !data3DHeader.sensorVendor.empty() )
+         scan.set( "sensorVendor", StringNode( imf_, data3DHeader.sensorVendor ) );
+      if ( !data3DHeader.sensorModel.empty() )
+         scan.set( "sensorModel", StringNode( imf_, data3DHeader.sensorModel ) );
+      if ( !data3DHeader.sensorSerialNumber.empty() )
+         scan.set( "sensorSerialNumber", StringNode( imf_, data3DHeader.sensorSerialNumber ) );
+      if ( !data3DHeader.sensorHardwareVersion.empty() )
+         scan.set( "sensorHardwareVersion", StringNode( imf_, data3DHeader.sensorHardwareVersion ) );
+      if ( !data3DHeader.sensorSoftwareVersion.empty() )
+         scan.set( "sensorSoftwareVersion", StringNode( imf_, data3DHeader.sensorSoftwareVersion ) );
+      if ( !data3DHeader.sensorFirmwareVersion.empty() )
+         scan.set( "sensorFirmwareVersion", StringNode( imf_, data3DHeader.sensorFirmwareVersion ) );
+
+      // Add temp/humidity to scan.
+      // Path names: "/data3D/0/temperature", etc...
+      if ( data3DHeader.temperature != E57_FLOAT_MAX )
+         scan.set( "temperature", FloatNode( imf_, data3DHeader.temperature ) );
+
+      if ( data3DHeader.relativeHumidity != E57_FLOAT_MAX )
+         scan.set( "relativeHumidity", FloatNode( imf_, data3DHeader.relativeHumidity ) );
+
+      if ( data3DHeader.atmosphericPressure != E57_FLOAT_MAX )
+         scan.set( "atmosphericPressure", FloatNode( imf_, data3DHeader.atmosphericPressure ) );
+
+      if ( ( data3DHeader.indexBounds.rowMinimum != 0 ) || ( data3DHeader.indexBounds.rowMaximum != 0 ) ||
+           ( data3DHeader.indexBounds.columnMinimum != 0 ) || ( data3DHeader.indexBounds.columnMaximum != 0 ) ||
+           ( data3DHeader.indexBounds.returnMinimum != 0 ) || ( data3DHeader.indexBounds.returnMaximum != 0 ) )
+      {
+         StructureNode ibox = StructureNode( imf_ );
+
+         //		if( (data3DHeader.indexBounds.rowMinimum != 0) ||
```

Comment:
Were comments these in the original? Do we need them?

Reply:
They came from the original. As far as I can tell it makes sense and it is allowed in the spec to write indexBounds just partially as guarded by those ifs, so I have uncommented them.